### DISCUSSION TITLE
🎉 Flash-DMA v0.1.0 - Initial Release

### DISCUSSION THREAD

**Discussion by LoserCheems** [SmallDoges/flash-dmattn#104] [Announcements]
We're excited to announce the first official release of Flash-DMA (Flash Dynamic Mask Attention)!

## 🚀 What is Flash-DMA?

Flash-DMA is a high-performance attention implementation that combines:
- **Flash Attention's memory efficiency** 
- **Dynamic Mask Attention's sparse computation**
- **Support for extremely long sequences** (128K+ tokens)

### 🔥 **Performance**
- **Sparse Attention**: Reduces computation from O(N²) to O(N·w) where w ≪ N
- **Memory Efficient**: Maintains O(N) memory complexity
- **CUDA Accelerated**: Custom sparse GEMM operations at kernel level

### 🛠️ **Multiple Backends**
- **CUDA Backend**: Maximum performance with custom kernels
- **Triton Backend**: Flexibility for research and development  
- **Flex Backend**: Integration with Transformers library

### 📏 **Long Sequence Support**
- Efficiently handles sequences of **128K+ tokens**
- Dynamic masking when sequence length exceeds `keep_window_size`
- Optimized memory layouts for large-scale processing

### Prerequisites
- Python 3.9+
- PyTorch 2.0+
- CUDA 11.8+
- NVIDIA GPU with Compute Capability 8.0+

### Install from Source
```bash
git clone https://github.com/SmallDoges/flash-dmattn.git
cd flash-dmattn
git submodule update --init --recursive
pip install .
```

## What's Changed
* Workspace by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/1
* Add namespace_config to csrc by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/2
* Add hardware_info to csrc by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/3
* Add block_info to csrc by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/4
* Add flash params to csrc by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/5
* Workspace by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/6
* Update golobal to Shared Memory operation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/7
* Update PREDICATES by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/8
* Fix some nits for layout by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/9
* Workspace by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/10
* Fix Dynamic Mask Attention Integration in FlashAttention CUDA Kernel by @Copilot in https://github.com/SmallDoges/flash-dmattn/pull/12
* Fix dynamic mask attention equivalence issue between Python and CUDA by @Copilot in https://github.com/SmallDoges/flash-dmattn/pull/14
* Fix CUDA dynamic mask attention scaling to match Python implementation by @Copilot in https://github.com/SmallDoges/flash-dmattn/pull/16
* Update mask.h by @Evanwu1125 in https://github.com/SmallDoges/flash-dmattn/pull/17
* Comprehensive README improvement with installation, usage examples, and documentation by @Copilot in https://github.com/SmallDoges/flash-dmattn/pull/19
* Adds no-topk variant for kernel performance analysis by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/20
* Optimize sparse GEMM and enable in attention computation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/21
* Adds row stride support to offset calculation methods by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/22
* Corrects ZOH tensor dimension comment by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/23
* Adds rotary positional encoding operations by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/24
* Adds conditional softcap switch macro by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/25
* Removes unused template parameter from DynamicMask by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/26
* Updates tensor offset calculations and formatting by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/27
* Adds split-K attention kernel with sparse computation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/28
* Adds dropout support and softcap feature to flash attention by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/29
* Add specialized CUDA kernels for multi-head attention with various head dimensions by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/30
* Remove cub submodule and add cutlass; implement FlashDynamicMaskAttention by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/31
* Refactors setup.py for production-ready package build by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/32
* Update integration by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/33
* Adds comprehensive API reference documentation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/34
* Updates README with improved technical accuracy and examples by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/35
* Fix bug by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/36
* Adds column stride parameters to ZOH_params struct by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/37
* Adds column stride support to offset calculations by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/38
* Fixes attention benchmarking and expands test coverage by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/39
* Reorders stride parameter assignments by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/40
* Adds column stride support to tensor memory layouts by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/41
* Improves code clarity and test coverage by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/42
* Updates copy function defaults and clarifies comments by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/43
* Updates copy operations to use improved vectorization by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/44
* Updates benchmark test configurations for better coverage by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/45
* Temporarily disables Split-KV feature by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/46
* Optimizes CUDA kernel block sizes for better occupancy by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/49
* Enables test case for 512x512 input dimensions by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/50
* Renames dzero_hold to dzoh and adds column stride by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/51
* Improves code formatting consistency in comments by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/52
* Fixes tensor addressing for ZOH and active mask in splitkv by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/53
* Refactor attention mask and bias structures for clarity by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/54
* Refactor backward kernel for attention mask and bias support by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/55
* Adds Flash Attention implementation with dynamic masking by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/56
* Fixes mask validation in forward kernel by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/57
* Fixes mask comparison and scaling logic in attention kernel by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/58
* Enhance Flash Attention with required parameters and improved backward pass by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/59
* Reorganizes flash attention files into instantiations directory by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/60
* Rename flash_dma to flash_dmattn and improve usability by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/61
* Adds bias gradient computation to backward kernel by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/62
* Add backend selection and dynamic mask attention support by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/63
* Update by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/64
* Removes no-topk CUDA implementation from benchmarks by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/65
* Enables comprehensive benchmark configurations by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/66
* Renames Flash Attention to SDPA in benchmark suite by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/67
* Refactors variable declarations for better readability by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/68
* Add bias gradient computation support in backward kernel by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/69
* Fix function naming and standardize memory copy alignment in attention kernel by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/70
* Adds unified mask application function with causal support by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/71
* Enables Split-KV avoidance and updates error messages by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/74
* Adds variable length forward pass support by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/75
* Simplify attention mask and bias parameter naming by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/76
* Remove unused parameters and simplify mask logic by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/77
* Add CUDA-integrated flash attention interface by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/78
* Improves version comparison using packaging library by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/79
* Refactor CUDA interface for improved usability by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/80
* Refactors CUDA implementation to use new interface by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/81
* Refactors dynamic mask function to improve clarity by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/82
* Expands API documentation with comprehensive interface guide by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/83
* Refactors API to use unified flash attention interface by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/84
* Convert banner image from JPG to PNG format by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/85
* Updates flash attention banner image by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/86
* Updates citation format and adds acknowledgment by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/87
* Adds comprehensive documentation with logo and Chinese translation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/89
* Simplify and standardize flex attention interface by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/90
* Standardize parameter naming and improve API consistency in attention functions by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/91
* Implement Doge model with dynamic attention mechanism by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/92
* Remove dropout functionality from flash attention by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/93
* Make attention parameters optional with defaults and simplify API documentation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/95
* Improves API parameter naming consistency by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/96
* Adds GitHub issue templates and PR template by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/97
* Streamlines benchmark suite structure and test scope by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/99
* Update contributor list, add citation metadata, and enhance documentation by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/100
* Remove paper citation and author information from README files by @LoserCheems in https://github.com/SmallDoges/flash-dmattn/pull/102

## New Contributors
* @LoserCheems made their first contribution in https://github.com/SmallDoges/flash-dmattn/pull/1
* @Copilot made their first contribution in https://github.com/SmallDoges/flash-dmattn/pull/12
* @Evanwu1125 made their first contribution in https://github.com/SmallDoges/flash-dmattn/pull/17

**Full Changelog**: https://github.com/SmallDoges/flash-dmattn/commits/0.1.0

<hr /><em>This discussion was created from the release <a href='https://github.com/SmallDoges/flash-dmattn/releases/tag/0.1.0'>🎉 Flash-DMA v0.1.0 - Initial Release</a>.</em>

**Comment by LoserCheems** [comment]
LGTM🤗
